apis/duck/typed.go

```diff
@@ -57,13 +57,43 @@ func (dif *TypedInformerFactory) Get(gvr schema.GroupVersionResource) (cache.Sha
 
 	lister := cache.NewGenericLister(inf.GetIndexer(), gvr.GroupResource())
 
-	go inf.Run(dif.StopChannel)
+	infChannel := make(chan struct{})
+	errorChannel := make(chan error)
 
-	if ok := cache.WaitForCacheSync(dif.StopChannel, inf.HasSynced); !ok {
-		return nil, nil, fmt.Errorf("failed starting shared index informer for %v with type %T", gvr, dif.Type)
-	}
+	go inf.Run(infChannel)
+
+	go func() {
+		if ok := cache.WaitForCacheSync(infChannel, inf.HasSynced); !ok {
+			errorChannel <- fmt.Errorf("failed starting shared index informer for %v with type %T", gvr, dif.Type)
+		} else {
+			errorChannel <- nil
+		}
+	}()
 
-	return inf, lister, nil
+	timer := time.NewTimer(time.Second * time.Duration(30))
```

Comment:
Add a comment that this specific time wasn't chosen for any particular reason, it just 'feels' right. And that it bounds how long a single threaded reconciler may be stuck processing this particular resource.